junit-vintage-engine/src/main/java/org/junit/vintage/engine/VintageTestEngine.java

```diff
@@ -70,17 +89,105 @@ public void execute(ExecutionRequest request) {
+		return parallelExecutionEnabled ? new ParallelExecutor(configurationParameters) : new SerialExecutor();
+	}
+
+	private void executeAllChildren(Executor executor, RunnerExecutor runnerExecutor, TestDescriptor engineDescriptor)
+			throws InterruptedException {
+		Collection<? extends TestDescriptor> children = engineDescriptor.getChildren();
+		CountDownLatch latch = new CountDownLatch(children.size());
 		// @formatter:off
-		engineDescriptor.getChildren()
+		children
 				.stream()
 				.map(RunnerTestDescriptor.class::cast)
-				.forEach(runnerExecutor::execute);
+				.map(descriptor -> (Runnable)() -> {
+					runnerExecutor.execute(descriptor);
+					latch.countDown();
+				})
+				.forEach(executor::execute);
 		// @formatter:on
+		latch.await();
+	}
+
+	/**
+	 * Wrapper for {@link Executor} to allow it to be used in a try-with-resources block.
+	 */
+	@API(status = INTERNAL, since = "5.8")
+	protected interface CloseableExecutor extends Executor, AutoCloseable {
+		default public void close() {
+		};
+	}
+
+	/**
+	 * {@link CloseableExecutor} that executes tasks synchronously.
+	 *
+	 * @since 5.8
+	 */
+	@API(status = INTERNAL, since = "5.8")
+	protected class SerialExecutor implements CloseableExecutor {
+		public void execute(Runnable command) {
+			command.run();
+		}
+
+	}
+
+	/**
+	 * {@link CloseableExecutor} backed by a {@link ForkJoinPool} that executes tasks asynchronously
+	 * based on the settings in {@value #PARALLEL_CONFIG}. Clients *must* implement their own logic
+	 * to wait for tasks to complete.
+	 *
+	 * @since 5.8
+	 */
+	@API(status = INTERNAL, since = "5.8")
+	protected class ParallelExecutor implements CloseableExecutor {
+
+		private final ExecutorService executorService;
+
+		/**
+		 * @param parameters test execution configuration
+		 */
+		public ParallelExecutor(ConfigurationParameters parameters) {
+			ConfigurationParameters prefixedParameters = new PrefixedConfigurationParameters(parameters,
+				PARALLEL_CONFIG);
+			String strategyName = prefixedParameters.get("strategy").orElse("dynamic");
+			ParallelExecutionConfigurationStrategy executionStrategy = DefaultParallelExecutionConfigurationStrategy.valueOf(
+				strategyName.toUpperCase());
+			ParallelExecutionConfiguration executionConfiguration = executionStrategy.createConfiguration(
+				prefixedParameters);
+			executorService = new ForkJoinPoolFactory().apply(executionConfiguration);
+		}
+
+		public void execute(Runnable command) {
+			executorService.execute(command);
+		}
+
+		public void close() {
+			executorService.shutdown();
+			Logger logger = LoggerFactory.getLogger(getClass());
+			try {
+				if (!executorService.awaitTermination(1, TimeUnit.MINUTES)) {
+					logger.warn(() -> "Timed out while waiting for tests to complete.");
+				}
+			}
+			catch (InterruptedException e) {
+				logger.warn(e, () -> "Interrupted while waiting for tests to complete.");
```

Comment:
```suggestion
				logger.warn(e, () -> "Interrupted while waiting for tests to complete.");
				Thread.currentThread().interrupt();
```

Reply:
I don't do this when waiting on the latch (line 97). Should I update that as well?